File: javascript/allergies/allergies.js
var allergies = [];

var Allergies = function (score) {
  this.allergicTo = function (allergy) {
    return allergies.indexOf(allergy) >= 0;
  };

  this.addAllergy = function (allergy, currentIndex, index) {
    if (currentIndex >= index) {
      if (!this.allergicTo(allergy)) {
        allergies.push(allergy);
      }
      while (currentIndex >= index) {
        currentIndex -= index;
      }
    }
    return currentIndex;
  };

  allergies = [];
  score = this.addAllergy("cats", score, 128);
  score = this.addAllergy("pollen", score, 64);
  score = this.addAllergy("chocolate", score, 32);
  score = this.addAllergy("tomatoes", score, 16);
  score = this.addAllergy("strawberries", score, 8);
  score = this.addAllergy("shellfish", score, 4);
  score = this.addAllergy("peanuts", score, 2);
  this.addAllergy("eggs", score, 1);


  this.list = function () {
    return allergies.reverse();
  };
};

module.exports = Allergies;

File: csharp/allergies/Allergies.cs
using System.Collections.Generic;

class Allergies
{
  private void AddAllergy(string allergy, ref int currentIndex, int index)
  {
    if (currentIndex >= index)
    {
      if (!AllergicTo(allergy))
        allergies.Add(allergy);
      while (currentIndex >= index)
        currentIndex -= index;
    }
  }

  public Allergies(int index)
  {
    allergies = new List<string>();
    AddAllergy("cats", ref index, 128);
    AddAllergy("pollen", ref index, 64);
    AddAllergy("chocolate", ref index, 32);
    AddAllergy("tomatoes", ref index, 16);
    AddAllergy("strawberries", ref index, 8);
    AddAllergy("shellfish", ref index, 4);
    AddAllergy("peanuts", ref index, 2);
    AddAllergy("eggs", ref index, 1);
  }

  public bool AllergicTo(string a_allergy)
  {
    return allergies.Exists(x => x == a_allergy);
  }

  public List<string> List()
  {
    var reverseList = allergies;
    reverseList.Reverse();
    return reverseList;
  }

  private List<string> allergies;
}